src/common/hist_util.cu

```diff
@@ -113,9 +112,9 @@ struct GPUSketcher {
 
    public:
     DeviceShard(int device, bst_uint row_begin, bst_uint row_end,
-                tree::TrainParam param) :
+                tree::TrainParam param, SketchContainer *sketch_container) :
```

Comment:
GPUSketcher already breaks SparsePage object into smaller batches. It should be possible to add another level of looping around that, so that quantile extraction with/without external memory is contained entirely within GPUSketcher.

Reply:
i have now made the quantile generation self contained within the gpu sketcher.